test/src/unittests/tonal/test_pitchsaliencefunction.py

```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env python
+
+# Copyright (C) 2006-2021  Music Technology Group - Universitat Pompeu Fabra
+#
+# This file is part of Essentia
+#
+# Essentia is free software: you can redistribute it and/or modify it under
+# the terms of the GNU Affero General Public License as published by the Free
+# Software Foundation (FSF), either version 3 of the License, or (at your
+# option) any later version.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the Affero GNU General Public License
+# version 3 along with this program. If not, see http://www.gnu.org/licenses/
+
+
+from numpy import *
+from essentia_test import *
+
+class TestPitchSalienceFunction(TestCase):
+  
+    def testInvalidParam(self):
+        self.assertConfigureFails(PitchSalienceFunction(), {'binResolution': -1})
+        self.assertConfigureFails(PitchSalienceFunction(), {'harmonicWeight': -1})
+        self.assertConfigureFails(PitchSalienceFunction(), {'harmonicWeight': 2})
+        self.assertConfigureFails(PitchSalienceFunction(), {'magnitudeCompression': -1})
+        self.assertConfigureFails(PitchSalienceFunction(), {'magnitudeCompression': 2})        
+        self.assertConfigureFails(PitchSalienceFunction(), {'magnitudeThreshold': -1})
+        self.assertConfigureFails(PitchSalienceFunction(), {'numberHarmonics': 0})
+        self.assertConfigureFails(PitchSalienceFunction(), {'numberHarmonics': -1})        
+        self.assertConfigureFails(PitchSalienceFunction(), {'referenceFrequency': -1})
+
+    def testZero(self):
+        freqs = zeros(1024)
+        mags  = zeros(1024)
+        self.assertRaises(RuntimeError, lambda: PitchSalienceFunction()(freqs, mags))
+
+    def testEmpty(self): 
+        self.assertEqualVector(PitchSalienceFunction()([], []), zeros(600))
+
+    def testOne(self):        
+        self.assertEqualVector(PitchSalienceFunction()(ones(1024), ones(1024)), zeros(600))
+
+    def testUnequalInputs(self):
+        # Choose a sample set of frequencies and magnitude vectors of unqual length
+        freqs = [250, 500, 1000, 2000, 3500] # length 5
+        mags = [0.5, 0.5, 0.5, 0.5] # length 4
+        self.assertRaises(EssentiaException, lambda: PitchSalienceFunction()(freqs, mags))
+
+    def testARealCase(self):
```

Comment:
>  The salience of a given frequency is computed as the sum of the weighted energies found at integer multiples (harmonics) of that frequency.

We are missing a test checking this. For example,
- Give a single peak as an input and validate how all subharmonics are activated in the salience function.
- Create a simple test with only a few input peaks. 

The [weights of harmonics](https://github.com/MTG/essentia/blob/master/src/algorithms/tonal/pitchsaliencefunction.cpp#L57) are defined by parameter `harmonicWeight`. 
- We can test the default 0.8 value.
- Test a simpler value of 1 and 0.

The values 0 and 1 aren't currently allowed for `harmonicWeight`, but we can enable those  as they may be useful in some practical use cases. They are also useful for running tests. We can run simple tests proposed in this comment  for both 0 and 1 values, apart from the default 0.8 value. Furthermore, the documentation already mentioned the case of `=1 for no decay` even though is not enabled.

Note that to make `harmonicWeight = 0` work, we need to adjust the code here: https://github.com/MTG/essentia/blob/master/src/algorithms/tonal/pitchsaliencefunction.cpp#L57
This is because the value of pow(0, 0) has undefined behavior. We should manually set the pow(0,0) value to 1. We can do it after the loop, setting `_harmonicWeights[0] = 1` in the case if `_harmonicWeight` == 0.

Reply:
I have addressed the other comments and will make a local commit on them.
This comment requires more work.

I will propose to close this PR and I split this PR into two parts:
1 part taking 3 algorithsm and other part taking 2 algorithms